### PR TITLE
Add benchmark for gRPC GetValidatorSet

### PR DESCRIPTION
**Pull request by hexfusion**
## Why this should be merged

It is important to understand the cost of requesting a large validator set over gRPC. Even though the call is local there is a cost for round trip serialization. The server + wire cost for serving 2000 recordsvis only about 1.6ms. The rest of the time is spent converting bytes to BLS keys (cgo) client side.

ref https://github.com/ava-labs/avalanchego/blob/7d73b59cb4838d304387ea680b9cc4053b72620c/snow/validators/gvalidators/validator_state_client.go#L75

```
$ go test -benchmem -run=^$ -bench ^BenchmarkGetValidatorSet$ github.com/ava-labs/avalanchego/snow/validators/gvalidators -memprofile benchvset.mem -cpuprofile benchvset.cpu

goos: linux
goarch: amd64
pkg: github.com/ava-labs/avalanchego/snow/validators/gvalidators
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkGetValidatorSet/get_validator_set_1_validators-12                 12841            106797 ns/op            9913 B/op        194 allocs/op
BenchmarkGetValidatorSet/get_validator_set_10_validators-12                 1710            670164 ns/op           16391 B/op        266 allocs/op
BenchmarkGetValidatorSet/get_validator_set_500_validators-12                  37          28670401 ns/op          386569 B/op       3744 allocs/op
BenchmarkGetValidatorSet/get_validator_set_1000_validators-12                 19          56178525 ns/op          763855 B/op       7269 allocs/op
BenchmarkGetValidatorSet/get_validator_set_2000_validators-12                  9         111610799 ns/op         1503906 B/op      14323 allocs/op
```
### profiles
[benchvset.cpu](https://github.com/ava-labs/avalanchego/files/11201944/benchvset.cpu.txt)
[benchvset.mem](https://github.com/ava-labs/avalanchego/files/11201959/benchvset.mem.txt)

## How this works

Uses golang native benchmark tooling

## How this was tested

manual

### CODE REVIEW

File: snow/validators/gvalidators/validator_state_test.go [hunk 1]
```diff
@@ -181,3 +182,49 @@ func TestGetValidatorSet(t *testing.T) {
 	_, err = state.client.GetValidatorSet(context.Background(), height, subnetID)
 	require.Error(err)
 }
+
+// BenchmarkGetValidatorSet measures the time it takes complete a gRPC client
```
Comment (darioush) [inline snow/validators/gvalidators/validator_state_test.go:186]:
```suggestion
// BenchmarkGetValidatorSet measures the time it takes to complete a gRPC client
```

File: snow/validators/gvalidators/validator_state_test.go [hunk 2]
```diff
@@ -181,3 +182,49 @@ func TestGetValidatorSet(t *testing.T) {
...
+	ctrl := gomock.NewController(b)
+	state := setupState(b, ctrl)
+	defer func() {
+		ctrl.Finish()
+		state.closeFn()
+	}()
+
+	height := uint64(1337)
+	subnetID := ids.GenerateTestID()
+	for i := 0; i < b.N; i++ {
```
Comment (darioush) [inline snow/validators/gvalidators/validator_state_test.go:208]: should we call `b.ResetTimer()` before this loop?

File: snow/validators/gvalidators/validator_state_test.go [hunk 3]
```diff
@@ -181,3 +182,49 @@ func TestGetValidatorSet(t *testing.T) {
...
+		_, err := state.client.GetValidatorSet(context.Background(), height, subnetID)
+		require.NoError(err)
+	}
+}
+
+func setupValidatorSet(b *testing.B, size int) map[ids.NodeID]*validators.GetValidatorOutput {
+	b.Helper()
+
+	set := make(map[ids.NodeID]*validators.GetValidatorOutput, size)
+	sk, _ := bls.NewSecretKey()
```
Comment (StephenButtolph) [inline snow/validators/gvalidators/validator_state_test.go:219]: We should either not take in `b`, or we should use it somewhere in this function... Generally I think it's good to handle the errors (even in tests) when possible
```suggestion
	sk, err := bls.NewSecretKey()
	require.NoError(b, err)
```

File: snow/validators/gvalidators/validator_state_test.go [hunk 4]
```diff
@@ -181,3 +182,49 @@ func TestGetValidatorSet(t *testing.T) {
...
+		state.closeFn()
+	}()
+
+	height := uint64(1337)
+	subnetID := ids.GenerateTestID()
+	for i := 0; i < b.N; i++ {
+		state.server.EXPECT().GetValidatorSet(gomock.Any(), height, subnetID).Return(vs, nil)
+		_, err := state.client.GetValidatorSet(context.Background(), height, subnetID)
+		require.NoError(err)
+	}
```
Comment (StephenButtolph) [inline snow/validators/gvalidators/validator_state_test.go:212]: I know it doesn't impact performance... but it seems weird that we are including the mocking and cleanup as part of the benchmark... (Also without actually running it, I wouldn't be able to say with any confidence that there isn't a performance impact... so I feel like this is just cleaner)
```suggestion
	state.server.EXPECT().GetValidatorSet(gomock.Any(), height, subnetID).Return(vs, nil).AnyTimes()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, err := state.client.GetValidatorSet(context.Background(), height, subnetID)
		require.NoError(err)
	}
	b.StopTimer()
```

File: snow/validators/gvalidators/validator_state_test.go [hunk 5]
```diff
@@ -181,3 +182,49 @@ func TestGetValidatorSet(t *testing.T) {
 	_, err = state.client.GetValidatorSet(context.Background(), height, subnetID)
 	require.Error(err)
 }
+
+// BenchmarkGetValidatorSet measures the time it takes to complete a gRPC client
+// request based on a mocked validator set.
+func BenchmarkGetValidatorSet(b *testing.B) {
+	for _, i := range []int{1, 10, 1000, 2000} {
+		vs := setupValidatorSet(b, i)
```
Comment (StephenButtolph) [inline snow/validators/gvalidators/validator_state_test.go:190]:
```suggestion
	for _, size := range []int{1, 10, 1000, 2000} {
		vs := setupValidatorSet(b, size)
```

**Review by StephenButtolph — COMMENTED** [review]
nits

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**